src/authn/index.js

```diff
@@ -97,15 +99,20 @@ function setup(app) {
         clientID: COGNITO_CLIENT_ID,
         callbackURL: "/logged-in",
         pkce: true,
-        state: true
+        state: true,
+        passReqToCallback: true,
       },
-      async (accessToken, refreshToken, {id_token: idToken}, profile, done) => {
-        // Verify both tokens for good measure, but pull user information from
-        // the identity token (which is its intended purpose).
+      async (req, accessToken, refreshToken, {id_token: idToken}, profile, done) => {
         try {
-          await verifyToken(accessToken, "access");
+          // Verifies tokens
+          const user = await userFromTokens({idToken, accessToken, refreshToken});
 
-          const user = await userFromIdToken(idToken);
+          // Only store tokens in session _after_ we verify them and have a user.
+          req.session.tokens = {
+            idToken,
+            accessToken,
+            refreshToken,
+          };
```

Comment:
Not naive! A good question that's important to ask.

The first thing to consider is the tokens themselves. What potential elevated access do they provide?

- On nextstrain.org the id token is used for authn, which may provide access to see private data or make changes to data.
- On Cognito's REST API for end users, these tokens don't provide any additional access (like an end user's ability to update their own user info) because the access token lacks the `aws.cognito.signin.user.admin` scope (note that this isn't "admin" in the sense of the whole user pool, just the single user).
- The tokens themselves embed some user info like email and name (and potentially phone, if set).
- No other services accept these tokens.

This is basically the same level of access that a signed nextstrain.org session id provides. The session ids in session data are unsigned though, so a compromise of session data wasn't directly actionable before (without also having our session secret), whereas the tokens in the session data are actionable without any additional secrets. It might be worth doing the work to store the secrets encrypted at rest now rather than later.

Our connection to Redis is secured by both authentication and TLS, but still, Redis listens on a public IP, which is completely unnecessary. This is an unfortunate property of Heroku's Redis add-on. We could switch to something else which would only be available internally, which would be nice, but probably means getting off Heroku to [Render](https://render.com) or [Fly](https://fly.io) (or running Redis ourselves as a sidecar dyno).

Not naive! A good question that's important to ask. The short answer: considerations, yes, issues, not likely.

The first thing to consider is the tokens themselves. What potential elevated access do they provide?

- On nextstrain.org the id token is used for authn, which may provide access to see private data or make changes to data.
- On Cognito's REST API for end users, these tokens don't provide any additional access (like an end user's ability to update their own user info) because the access token lacks the `aws.cognito.signin.user.admin` scope (note that this isn't "admin" in the sense of the whole user pool, just the single user).
- The tokens themselves embed some user info like email and name (and potentially phone, if set).

This is basically the same level of access that a signed nextstrain.org session id provides. The session ids in session data are unsigned though, so a compromise of session data wasn't directly actionable before (without also having our session secret), whereas the tokens in the session data are actionable without any additional secrets.

Our connection to Redis is secured by both authentication and TLS (without hostname verification :cry:), though Redis still listens on a public IP, which is completely unnecessary. This is an unfortunate property of Heroku's Redis add-on. We could switch to something else which would only be available internally, which would be nice, but probably means getting off Heroku to [Render](https://render.com) or [Fly](https://fly.io) (or running Redis ourselves as a sidecar dyno).

Reply:
With my latest push, I added encryption at rest and documented security risks/protections around sessions in a new doc page. Thanks again for the nudge to think a little deeper here!